src/command/info/info.c

```diff
@@ -350,6 +358,271 @@ backupList(VariantList *backupSection, InfoBackup *info, const String *backupLab
+            KeyValue *backupDbInfo = varKv(kvGet(backupInfo, KEY_DATABASE_VAR));
+            unsigned int backupDbId = varUInt(kvGet(backupDbInfo, DB_KEY_ID_VAR));
+
+            if (backupDbId == dbId)
+            {
+                KeyValue *archiveBackupInfo = varKv(kvGet(backupInfo, KEY_ARCHIVE_VAR));
+
+                if (kvGet(archiveBackupInfo, KEY_START_VAR) != NULL &&
+                    kvGet(archiveBackupInfo, KEY_STOP_VAR) != NULL)
+                {
+                    // Generate the list of required WAL archives for each backup
+                    generateNeededArchivesList(
+                        neededArchivesList, pgVersion, stanza, archiveId,
+                        varStr(kvGet(archiveBackupInfo, KEY_START_VAR)),
+                        varStr(kvGet(archiveBackupInfo, KEY_STOP_VAR)));
+
+                    // Store the first backup start WAL location
+                    if (firstStartWal == NULL)
+                        firstStartWal = varStrForce(kvGet(archiveBackupInfo, KEY_START_VAR));
+
+                    if(strEq(varStr(kvGet(backupInfo, BACKUP_KEY_TYPE_VAR)), BACKUP_TYPE_FULL_STR))
+                    {
+                        strLstAdd(backupStartWalsList, varStr(kvGet(archiveBackupInfo, KEY_START_VAR)));
+                    }
+                    else if(strEq(varStr(kvGet(backupInfo, BACKUP_KEY_TYPE_VAR)), BACKUP_TYPE_DIFF_STR) && (
+                            strEq(cfgOptionStr(cfgOptRepoRetentionArchiveType), BACKUP_TYPE_DIFF_STR) ||
+                            strEq(cfgOptionStr(cfgOptRepoRetentionArchiveType), BACKUP_TYPE_INCR_STR)))
+                    {
+                        strLstAdd(backupStartWalsList, varStr(kvGet(archiveBackupInfo, KEY_START_VAR)));
+                    }
+                    else if(strEq(varStr(kvGet(backupInfo, BACKUP_KEY_TYPE_VAR)), BACKUP_TYPE_INCR_STR) &&
+                            strEq(cfgOptionStr(cfgOptRepoRetentionArchiveType), BACKUP_TYPE_INCR_STR))
+                    {
+                        strLstAdd(backupStartWalsList, varStr(kvGet(archiveBackupInfo, KEY_START_VAR)));
+                    }
+                }
+            }
+        }
+
+        // Check that the oldest archive in the repo is not older than the first backup start WAL location.
+        // It could indicate an expire problem.
+        if (strCmp(archiveStart,firstStartWal) < 0)
+            LOG_WARN_FMT(
+                "min wal archive ('%s') found in the repo is older than the first backup start WAL location ('%s').\n"
+                "HINT: this might indicate a configuration error in the retention policy.",
+                strZ(archiveStart), strZ(firstStartWal));
+
+        // If repo-retention-archive is set, exclude potentially expired archives
+        if (cfgOptionTest(cfgOptRepoRetentionArchive))
+        {
+            archiveStart = strLstGet(backupStartWalsList,
+                strLstSize(backupStartWalsList) - cfgOptionUInt(cfgOptRepoRetentionArchive));
```

Comment:
I'm not clear on the backupStartWalList - if we take the example I gave above, then by the time we get here, backupStartWalsList would contain 000000010000000000000002 (FULL start), 000000010000000000000003 and 000000010000000000000004 (start from the 2 DIFFs). If archive-retention-type=DIFF and archive-retention=4 then this code would error (3-4 = -1). Even if archive-retention were 2 instead, the archiveStart would be backupStartWalsList[1] = 000000010000000000000003 which would only mean the neededArchiveList would start at 000000010000000000000002 instead of 000000010000000000000001.

Also, this does not take into account time-based backup retention or adhoc backup expiration. 

So what is the real goal here? Are we trying to inform users that there are backups that have gaps in the WAL between the backup wal start/stop, meaning that backup is invalid? Or that there are gaps between the backups so that they can't perform PITR? 

I know you are writing some test, so maybe it will be clearer to me once there are some examples of the expected results.

Reply:
With archive-retention=4, we'd have a WARNing saying we keep more archives than backups.

My idea is to inform if there are missing WAL archives between start/stop backup location but also between backups themselves. But I'd also like to take care correctly of the retention policy, to avoid reporting missing archives where they're not missing since they've been expired. I should verify that part to also take care of the time-based retention.

You're right, it should be easier to see with some tests. I'll try to submit them asap.